File: app/src/main/java/life/plank/juna/zone/view/adapter/fixture/FixtureAdapter.java
package life.plank.juna.zone.view.adapter.fixture;

import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.request.RequestOptions;
import com.github.mikephil.charting.charts.LineChart;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import life.plank.juna.zone.R;
import life.plank.juna.zone.data.model.MatchFixture;
import life.plank.juna.zone.interfaces.LeagueContainer;
import life.plank.juna.zone.util.view.BaseRecyclerView;

import static life.plank.juna.zone.util.common.AppConstants.MatchTimeVal.MATCH_LIVE;
import static life.plank.juna.zone.util.common.DataUtil.getSeparator;
import static life.plank.juna.zone.util.common.ScrubberUtilKt.loadScrubber;
import static life.plank.juna.zone.util.time.DateUtil.getMatchTimeValue;
import static life.plank.juna.zone.util.view.UIDisplayUtil.alternateBackgroundColor;
import static life.plank.juna.zone.util.view.UIDisplayUtil.getDp;
import static life.plank.juna.zone.util.view.UIDisplayUtil.getEndDrawableTarget;
import static life.plank.juna.zone.util.view.UIDisplayUtil.getStartDrawableTarget;

public class FixtureAdapter extends BaseRecyclerView.Adapter<BaseRecyclerView.ViewHolder> {

    private static final int FIXTURE_LIVE = 0;
    private static final int FIXTURE_NOT_LIVE = 1;

    private List<MatchFixture> matchFixtureList;
    private LeagueContainer leagueContainer;

    FixtureAdapter(List<MatchFixture> matchFixtureList, LeagueContainer leagueContainer) {
        this.matchFixtureList = matchFixtureList;
        this.leagueContainer = leagueContainer;
        if (this.matchFixtureList == null) {
            this.matchFixtureList = new ArrayList<>();
        }
    }

    @NonNull
    @Override
    public BaseRecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == FIXTURE_LIVE) {
            return new LiveFixtureViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_fixture_league, parent, false), this);
        } else {
            return new NotLiveFixtureViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_fixture, parent, false), this);
        }
    }

    @Override
    public int getItemCount() {
        return matchFixtureList.size();
    }

    @Override
    public int getItemViewType(int position) {
        return getMatchTimeValue(matchFixtureList.get(position).getMatchStartTime(), false) == MATCH_LIVE ? FIXTURE_LIVE : FIXTURE_NOT_LIVE;
    }

    public void update(List<MatchFixture> matchFixtureList) {
        this.matchFixtureList = matchFixtureList;
        notifyDataSetChanged();
    }

    public static class NotLiveFixtureViewHolder extends BaseRecyclerView.ViewHolder {

        @BindView(R.id.home_team_name)
        TextView homeTeamName;
        @BindView(R.id.home_team_logo)
        ImageView homeTeamLogo;
        @BindView(R.id.date_schedule)
        TextView separatorView;
        @BindView(R.id.win_pointer)
        ImageView winPointer;
        @BindView(R.id.visiting_team_logo)
        ImageView visitingTeamLogo;
        @BindView(R.id.visiting_team_name)
        TextView visitingTeamName;

        private final WeakReference<FixtureAdapter> ref;
        private MatchFixture scoreFixture;

        NotLiveFixtureViewHolder(View itemView, FixtureAdapter adapter) {
            super(itemView);
            this.ref = new WeakReference<>(adapter);
            ButterKnife.bind(this, itemView);
        }

        @Override
        public void bind() {
            scoreFixture = ref.get().matchFixtureList.get(getAdapterPosition());
            alternateBackgroundColor(itemView, getAdapterPosition());

            ref.get().leagueContainer.getGlide()
                    .load(scoreFixture.getHomeTeam().getLogoLink())
                    .apply(RequestOptions.centerCropTransform().placeholder(R.drawable.ic_place_holder).error(R.drawable.ic_place_holder))
                    .into(homeTeamLogo);

            ref.get().leagueContainer.getGlide()
                    .load(scoreFixture.getAwayTeam().getLogoLink())
                    .apply(RequestOptions.centerCropTransform().placeholder(R.drawable.ic_place_holder).error(R.drawable.ic_place_holder))
                    .into(visitingTeamLogo);

            homeTeamName.setText(scoreFixture.getHomeTeam().getName());
            visitingTeamName.setText(scoreFixture.getAwayTeam().getName());
            separatorView.setText(getSeparator(scoreFixture.toMatchDetails(), winPointer, false));
        }

        @OnClick(R.id.root_layout)
        void onFixtureClicked() {
            ref.get().leagueContainer.onFixtureSelected(scoreFixture, ref.get().leagueContainer.getTheLeague());
        }
    }

    static class LiveFixtureViewHolder extends BaseRecyclerView.ViewHolder {

        @BindView(R.id.home_team_name)
        TextView homeTeamNameTextView;
        @BindView(R.id.score)
        TextView scoreTextView;
        @BindView(R.id.win_pointer)
        ImageView winPointer;
        @BindView(R.id.visiting_team_name)
        TextView visitingTeamNameTextView;
        @BindView(R.id.time_status)
        TextView timeStatusTextView;
        @BindView(R.id.scrubber)
        LineChart scrubber;

        private final WeakReference<FixtureAdapter> ref;
        private MatchFixture scoreFixture;

        LiveFixtureViewHolder(View itemView, FixtureAdapter adapter) {
            super(itemView);
            ref = new WeakReference<>(adapter);
            ButterKnife.bind(this, itemView);
        }

        @Override
        public void bind() {
            scoreFixture = ref.get().matchFixtureList.get(getAdapterPosition());
            alternateBackgroundColor(itemView, getAdapterPosition());

            ref.get().leagueContainer.getGlide()
                    .load(scoreFixture.getHomeTeam().getLogoLink())
                    .apply(RequestOptions.centerCropTransform()
                            .placeholder(R.drawable.ic_place_holder)
                            .error(R.drawable.ic_place_holder)
                            .override((int) getDp(12), (int) getDp(12)))
                    .into(getEndDrawableTarget(homeTeamNameTextView));

            ref.get().leagueContainer.getGlide()
                    .load(scoreFixture.getAwayTeam().getLogoLink())
                    .apply(RequestOptions.centerCropTransform()
                            .placeholder(R.drawable.ic_place_holder)
                            .error(R.drawable.ic_place_holder)
                            .override((int) getDp(12), (int) getDp(12)))
                    .into(getStartDrawableTarget(visitingTeamNameTextView));

            homeTeamNameTextView.setText(scoreFixture.getHomeTeam().getName());
            visitingTeamNameTextView.setText(scoreFixture.getAwayTeam().getName());
            scoreTextView.setText(getSeparator(scoreFixture.toMatchDetails(), winPointer, false));
            timeStatusTextView.setText(scoreFixture.getTimeStatus());
            loadScrubber(scrubber, false);
        }

        @OnClick(R.id.root_layout)
        void onFixtureClicked() {
            ref.get().leagueContainer.onFixtureSelected(scoreFixture, ref.get().leagueContainer.getTheLeague());
        }
    }
}

File: app/src/main/java/life/plank/juna/zone/view/adapter/fixture/FixtureAdapter.kt
package life.plank.juna.zone.view.adapter.fixture

import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.SimpleTarget
import kotlinx.android.synthetic.main.item_fixture.view.*
import kotlinx.android.synthetic.main.item_fixture_league.view.*
import life.plank.juna.zone.R
import life.plank.juna.zone.data.model.MatchFixture
import life.plank.juna.zone.interfaces.LeagueContainer
import life.plank.juna.zone.util.common.AppConstants.MatchTimeVal.MATCH_LIVE
import life.plank.juna.zone.util.common.DataUtil.getSeparator
import life.plank.juna.zone.util.common.loadScrubber
import life.plank.juna.zone.util.facilis.onDebouncingClick
import life.plank.juna.zone.util.time.DateUtil.getMatchTimeValue
import life.plank.juna.zone.util.view.UIDisplayUtil.*
import java.util.*

class FixtureAdapter(
        private var matchFixtureList: List<MatchFixture> = ArrayList(),
        private val leagueContainer: LeagueContainer
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val FIXTURE_LIVE = 0
        private const val FIXTURE_NOT_LIVE = 1
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            FIXTURE_LIVE -> LiveFixtureViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_fixture_league, parent, false))
            else -> NotLiveFixtureViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_fixture, parent, false))
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val scoreFixture = matchFixtureList[position]
        when (holder.itemViewType) {
            FIXTURE_LIVE -> {
                holder.itemView.run {
                    alternateBackgroundColor(this, position)

                    leagueContainer.getGlide()
                            .load(scoreFixture.homeTeam.logoLink)
                            .apply(RequestOptions.centerCropTransform()
                                    .placeholder(R.drawable.shimmer_circle)
                                    .error(R.drawable.ic_place_holder)
                                    .override(getDp(12f).toInt(), getDp(12f).toInt()))
                            .into<SimpleTarget<Drawable>>(getEndDrawableTarget(live_home_team_name))

                    leagueContainer.getGlide()
                            .load(scoreFixture.awayTeam.logoLink)
                            .apply(RequestOptions.centerCropTransform()
                                    .placeholder(R.drawable.shimmer_circle)
                                    .error(R.drawable.ic_place_holder)
                                    .override(getDp(12f).toInt(), getDp(12f).toInt()))
                            .into<SimpleTarget<Drawable>>(getStartDrawableTarget(live_visiting_team_name))

                    live_home_team_name.text = scoreFixture.homeTeam.name
                    live_visiting_team_name.text = scoreFixture.awayTeam.name
                    live_score.text = getSeparator(scoreFixture.toMatchDetails(), win_pointer, false)
                    live_time_status.text = scoreFixture.timeStatus

                    loadScrubber(live_scrubber, false)
                }
            }
            else -> {
                holder.itemView.run {
                    alternateBackgroundColor(this, position)

                    leagueContainer.getGlide()
                            .load(scoreFixture.homeTeam.logoLink)
                            .apply(RequestOptions.centerCropTransform().placeholder(R.drawable.shimmer_circle).error(R.drawable.ic_place_holder))
                            .into(home_team_logo)

                    leagueContainer.getGlide()
                            .load(scoreFixture.awayTeam.logoLink)
                            .apply(RequestOptions.centerCropTransform().placeholder(R.drawable.shimmer_circle).error(R.drawable.ic_place_holder))
                            .into(visiting_team_logo)

                    home_team_name.text = scoreFixture.homeTeam.name
                    visiting_team_name.text = scoreFixture.awayTeam.name
                    date_schedule.text = getSeparator(scoreFixture.toMatchDetails(), win_pointer, false)

                }
            }
        }
        holder.itemView.onDebouncingClick { leagueContainer.onFixtureSelected(scoreFixture, leagueContainer.getTheLeague()) }
    }

    override fun getItemCount(): Int = matchFixtureList.size

    override fun getItemViewType(position: Int): Int = if (getMatchTimeValue(matchFixtureList[position].matchStartTime, false) == MATCH_LIVE) FIXTURE_LIVE else FIXTURE_NOT_LIVE

    fun update(matchFixtureList: List<MatchFixture>) {
        this.matchFixtureList = matchFixtureList
        notifyDataSetChanged()
    }

    class NotLiveFixtureViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    class LiveFixtureViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
}